### PR TITLE
Fixes #26: Adds support for handling apostrophes in utterances 

### LINKED ISSUE #26: Intent does not handle apostrophes properly
Give the following intent
```
add {Food} to (| (the | my)) {ShoppingList} (| list) (under {Category} |)
```
Shopping list is improperly parsed
 Using the phrase:
`add temperature sensors to steve's projecta`

The intent parser produces
`steve ' s projecta` instead of `steve's projecta`

The utterance shows correct parsing:
```
~~~~50788 | __main__:handle_utterance:72 | Utterance: ["add temperature sensors to steve's projects"]
```

however the `message.data` shows that `shoppinglist` has been poorly parsed

```
 {'food': 'temperature sensors', 'shoppinglist': "steve ' s projects", 'utterance': "add temperature sensors to steve's projects"}

```
This obviously causes a errors or unmatched entities

The error comes from match_data.py. This statement:
```
def detokenize(self):
        self.sent = ' '.join(self.sent)
```
combine with the fact that `self.sent` is split like so:
```
 'sent': ['add', 'something', 'to', 'steve', "'", 's', 'projects'], 'matches': {}, 'conf': 0.0}
```
Causes the error. One solution that could be refined is

```
    @staticmethod
    def handle_apostrophes(old_sentence):
        new_sentence = ''
        apostrophe_present = False
        sentence = "steve's projects"

        for word in old_sentence:
          if word == "'":
              apostrophe_present = True
              new_sentence += word
          else:
            if apostrophe_present:
                new_sentence += word
                apostrophe_present = False
            else:
                if len(new_sentence) > 0:
                    new_sentence += " " + word
                else:
                    new_sentence = word
            apostrophe_presnet = False
        return new_sentence  


    # Converts parameters from lists of tokens to one combined string
    def detokenize(self):
        self.sent = self.handle_apostrophes(self.sent)

        new_matches = {}
        for token, sent in self.matches.items():
            print(self.handle_apostrophes(sent))
            new_token = token.replace('{', '').replace('}', '')
            new_matches[new_token] = self.handle_apostrophes(sent)
        self.matches = new_matches
```

### PR DESCRIPTION
**Pull request by stratus-ss**
This fixes issue #26 

 adds a static method for handling a sentence and returns a properly constructed sentence from a list object

### CODE REVIEW

**Comment by forslund** [comment]
Hi @stratus-ss thanks for contributing this fix! Do you think you could add a test case as well?

Also before we can merge the PR we'll need you to sign a [contributor license agreement](https://mycroft.ai/CLA). This is to protect yourself, the project, and users of Mycroft technologies granting Mycroft AI to License your work free of charge while you retain all ownership of the code you contributed.

Once more thank you for fixing this!

**Comment by stratus-ss** [comment]
I put in the request for the CLA. In the mean time, do you have a standard/documentation for desired test cases? 

I haven't started digging into VK yet so any pointers would be appreciated

**Comment by forslund** [comment]
For this I mean a unittest check the test folder in this repo for existing tests as examples.Perhaps add a test to the test_intent.py file

**Comment by stratus-ss** [comment]
would it be better to add a test to match_data.py since it already seems to be testing the same file?

**Comment by forslund** [comment]
Ah yes, that's a much better place. Sorry for my confusion.

**Comment by stratus-ss** [comment]
CLA has been signed

Test has been updated to include this

```
class TestMatchData:
    def setup(self):
        self.sentence = ["it", "'", "s", "a", "new", "sentence"]

    def test_handle_apostrophes(self):
        joined_sentence = self.match.handle_apostrophes(self.sentence)
        assert joined_sentence == "it's a new sentence"
```

This should be ready for review

**Comment by krisgesling** [comment]
CLA received, thanks Stratus-ss. 

Will leave the review to @forslund

File: padatious/match_data.py [hunk 1]
```diff
@@ -42,11 +42,38 @@ def get(self, key, default=None):
...
+        return new_sentence
+
     # Converts parameters from lists of tokens to one combined string
     def detokenize(self):
-        self.sent = ' '.join(self.sent)
+        self.sent = self.handle_apostrophes(self.sent)
+
         new_matches = {}
         for token, sent in self.matches.items():
+            print(self.handle_apostrophes(sent))
```
Comment (forslund) [inline padatious/match_data.py]: Should this print be here or is this for debugging?

**Review by forslund — COMMENTED** [review]
My English isn't the best. Would this work correctly for possessive plurals like "the twins' parents"?

File: padatious/match_data.py (hunk 1 above)
Comment (stratus-ss) [reply padatious/match_data.py]: good catch! I'll remove that

**Comment by stratus-ss** [comment]
> My English isn't the best. Would this work correctly for possessive plurals like "the twins' parents"?

I don't think so. I haven't been able to get mycroft to produce this gramatic case. I'm not sure how to test for this programmatically since it comes in as a list. Without extensive testing, I am not sure how you can (from a list) determine what is ["steve", "'", "s"] and ["twins", "'", "parent"]

Maybe checking the length of the next index is a way?

**Comment by forslund** [comment]
True it's hard and the rules for this will vary between languages

**Comment by stratus-ss** [comment]
> True it's hard and the rules for this will vary between languages

updated the code and the test to handle possessives with 's' + apostrophe such as "the parents' house"

File: tests/test_match_data.py [hunk 2]
```diff
@@ -18,10 +18,17 @@
...
         self.match.detokenize()
         assert self.match.sent == 'one two'
 
         correct_match = MatchData('name', 'one two', {'word': 'value tokens'}, 0.5)
         assert self.match.__dict__ == correct_match.__dict__
+
+    def test_handle_apostrophes(self):
+        joined_sentence = self.match.handle_apostrophes(self.sentence)
+        assert joined_sentence == "it's a new sentence"
+        assert joined_sentence == "the parents' house"
```
Comment (forslund) [inline tests/test_match_data.py]: I think before this line there should be a line

```
joined_sentence = self.match.handle_apostrophes(self.sentence2)
```

**Review by forslund — COMMENTED** [review]
Thanks @stratus-ss I think it's working as intended but the test case seems to be missing a line.

File: tests/test_match_data.py (hunk 2 above)
Comment (stratus-ss) [reply tests/test_match_data.py]: added a second variable. You're right I missed that. Thanks

**Comment by forslund** [comment]
Excellent! Works as intended. If you don't mind I'll rebase it slightly to clean up the commit history before merging it.

**Comment by stratus-ss** [comment]
> Excellent! Works as intended. If you don't mind I'll rebase it slightly to clean up the commit history before merging it.

ya go ahead and squash those commits, there is a lot of garbage in there

**Comment by forslund** [comment]
Merged, thank you very much for this contribution!

**Pull request merged**